### PR TITLE
#72: Added method generate bucket udf path

### LINKED ISSUE #72: Add method generate_bucket_udf_path to BucketFSLocation
# Background

- We use the two implementations BucketFSLocation together with the BucketFSFactory in our UDFs, where we get the BucketFS location url injected via a connection object
- Depending on the url we either get the LocalFSMockBucketFSLocation or BucketFSLocation
- Often we need also the path in the localfs for the bucketfs location
- However, until now, we only could get it via bucketfs_utils.generate_bucket_udf_path which required a BucketConfig which we don't have for the LocalFSMockBucketFSLocation
- For that reason, we need to add the [generate_bucket_udf_path](https://github.com/exasol/bucketfs-utils-python/blob/a0455eb435bcc42459898bd95b007748f1ceb13c/exasol_bucketfs_utils_python/bucketfs_utils.py#L50) method to both BucketFSLocation implementation

# Acceptance Criteria

- Add method generate_bucket_udf_path to BucketFSLocation interface and implementation
- Add tests

### PR DESCRIPTION
**Pull request by umitbuyuksahin**
Fixes #72

### CODE REVIEW

File: exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py [hunk 1]
```diff
@@ -20,6 +22,20 @@ def __init__(self, base_path: PurePosixPath):
...
 
+    def generate_bucket_udf_path(
+            self, path_in_bucket: Union[None, str, PurePosixPath]) \
+            -> PurePosixPath:
+
+        if path_in_bucket is not None:
+            path_in_bucket = bucketfs_utils.\
+                make_path_relative(path_in_bucket)
+            path_in_bucket = bucketfs_utils.\
+                correct_path_in_bucket_for_archives(path_in_bucket)
```
Comment (tkilias) [inline exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py:33]: this is one of the major points where the mock and the real thing differ. The real bucketfs extract archives files, such that they are available as directories in the UDFs. However, this we don't support in the mock. So, we need to think about this correction.

File: tests/test_bucketfs_location.py [hunk 2]
```diff
@@ -1,16 +1,36 @@
...
 from exasol_bucketfs_utils_python.bucketfs_connection_config import BucketFSConnectionConfig
-
 from exasol_bucketfs_utils_python.bucketfs_location import BucketFSLocation
 import pytest
 import textwrap
 from tests.test_load_fs_file_from_udf import delete_testfile_from_bucketfs, upload_testfile_to_bucketfs
 # TODO replace upload_testfile_to_BucketFS once missing funcs in BucketFSLocation are implemented
 
 
+@pytest.mark.parametrize("path_in_bucket", [
```
Comment (tkilias) [inline tests/test_bucketfs_location.py:12]: I think, we should also have a integration test with the db, which verifies, that we find the path we are generating

**Review by tkilias — CHANGES_REQUESTED** [review]

File: exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py (hunk 1 above)
Comment (umitbuyuksahin) [reply exasol_bucketfs_utils_python/localfs_mock_bucketfs_location.py:33]: We are uploading the archive files directly now, then we have to remove this control. the question is that should we provide extracting archive feature in localfs?

File: tests/test_bucketfs_location.py [hunk 3]
```diff
@@ -1,16 +1,105 @@
...
+    bucketfs_location = BucketFSLocation(bucket_config, "")
+
+    udf_path = bucketfs_location.generate_bucket_udf_path(
+        path_in_bucket=path_in_bucket)
+
+    assert str(udf_path) == "/buckets/bfsdefault/default/" \
+                            "path/in/bucket/file.txt"
+
+
+def test_generate_bucket_udf_path_with_db(
```
Comment (tkilias) [inline tests/test_bucketfs_location.py:35]: Please, Split unit tests ans integration tests into seperaten files and directories, this allows easier selection which  tests to run

File: tests/test_bucketfs_location.py [hunk 4]
```diff
@@ -1,16 +1,105 @@
...
+                return PurePosixPath("/buckets", bucketfs_name, bucket_name)
+                
+            def run(ctx):
+                path_in_bucket = ctx.path_in_bucket
+                bucket_config = get_bucket_config()
+                bucket_path = get_bucket_path()
+                bucketfs_location = BucketFSLocation(bucket_config, "")
+                relative_file_path = bucketfs_location.generate_bucket_udf_path(
+                    path_in_bucket).relative_to(bucket_path)
+                listed_files = bucketfs_location.list_files_in_bucketfs(".")
```
Comment (tkilias) [inline tests/test_bucketfs_location.py:89]: I meant with check if the files are there, that you check if the file exists in local filesystem at the path in the UDF, which you got by function you implemented.

bucketfs_location.generate_bucket_udf_path, path_in_bucket).exists()

Basically,

**Review by tkilias — CHANGES_REQUESTED** [review]

**Review by tkilias — APPROVED** [review]

**Pull request merged**